cmd/generate.go

```diff
@@ -247,97 +255,161 @@ function fetch_bench_io_results() {
+# in the client node.
+function modify_remote_hosts_on_client_node() {
+  # client_node is the one to run TCP_RR and TCP_STREAM.
+  local client_node=$1
+  local server_node=$2
+  # test_mode should be either cross-region or intra-az.
+  local test_mode=$3
+
+  local server_ip=$(roachprod ip "$server_node")
+  if [ -z $server_ip ]
   then
-    node="$CLUSTER":1
-  else
-    node="$CLUSTER":3
+    echo "cannot get server_ip FOR server (remote) node ($server_node) in network test"
+    exit 1
   fi
-  roachprod run $node ./scripts/gen/network-netperf.sh -- -w
-  copy_result_with_retry $node "netperf-results"
+  
+  # Since linux doesn't allow ":" in filename, we replace the ":" in 
+  # $server_node to "-".
+  formatted_server_node=$(echo "${server_node//:/-}")
+  echo "formatted_server_node=$formatted_server_node"
+  
+  # Save the ip address of the server node into the 
+  # netperf/doc/examples/remote_hosts in the client node.
+  local remote_host_file="${logdir}/${formatted_server_node}_${test_mode}_remote_hosts"
+  printf "REMOTE_HOSTS[0]=$server_ip\nREMOTE_HOSTS[1]=$server_ip\nNUM_REMOTE_HOSTS=2\n" >"$remote_host_file"
+  chmod 777 "$remote_host_file"
+  roachprod run "$client_node" -- sudo chmod 777 -R netperf
+  roachprod put "$client_node" "$remote_host_file" netperf/doc/examples/${test_mode}_remote_hosts
 }
 
-# Run TPCC Benchmark
-function bench_tpcc() {
+# get_best_number_streams is to run a netperf TCP_STREAM test with 
+# gradually incrementing the number of streams until the aggregate throughput 
+# converges (the latest 3 agg throughput's std < 0.3). The best number of 
+# streams will be saved in a file "num_streams" in the client node.
+function get_best_number_streams() {
+  local client_node=$1
+  local test_mode=$2
+  echo "running getting best num of stream for $client_node"
+
+  roachprod run "$client_node" -- "cd netperf/doc/examples && SEARCH_BEST_NUM_STREAMS=1 TEST_MODE=$test_mode ./runemomniaggdemo.sh"
+  echo "get best number of stream for $client_node"
+}
+
+# run_netperf_between_server_client is to get the best number of streams 
+# to run the throughput test between the server and client node, and run the
+# netperf latency and throughput test. We start the netserver on the server node
+# and run netperf command on the client node.
+# Note that in the cross-region case, we set the east node as the client node, 
+# and the west node as the server node. 
+function run_netperf_between_server_client() {
+  
+  local client_node=$1
+  local server_node=$2
+
+  local PORT=$3
+  local test_mode=$4
+  local netperf_extra_args=$5
+
+  local server_ip=$(roachprod ip $server_node)
+
+  roachprod run $client_node sudo ./scripts/gen/network-setup.sh
+  roachprod run $server_node sudo ./scripts/gen/network-setup.sh
+
+  # Start netserver on the server node.
+  # It may give error, but it only means that the netserver is already running
+  # on the given port, so we should proceed when that happens.
+  set +x
```

Comment:
the whole script runs with +x.  I don't think you need this.

Reply:
Actually, the whole script is run with `set -ex`, which means the running script will exit once any command hits an error.  I'm using `set +x ... set -x` to make an "exception block" in this script -- in this block, if any command hits an error, we proceed with the script rather than exit immediately. See also https://stackoverflow.com/a/70431094/10400141

Reply:
Okay... I missed that part... But if you only worried about server running perhaps you can test for that?
lsof? Or maybe there is a flag to netperf to ignore runnign servers?
Anyways, this is fine by me -- which ever way it goes.

Reply:
> I'm using set +x ... set -x to make an "exception block" in this script 

Did you mean `set -e` instead? `set [+-] x` toggles trace mode which echoes each executed command.

Reply:
Ah yes! Made the changes.